python/ray/rllib/evaluation/sampler.py

```diff
@@ -262,184 +262,242 @@ def new_episode():
         unfiltered_obs, rewards, dones, infos, off_policy_actions = \
             async_vector_env.poll()
 
-        # Map of policy_id to list of PolicyEvalData
-        to_eval = defaultdict(list)
+        # Process observations and prepare for policy evaluation
+        active_envs, to_eval, outputs = _process_observations(
+            async_vector_env, policies, batch_builder_pool, active_episodes,
+            unfiltered_obs, rewards, dones, infos, off_policy_actions, horizon,
+            obs_filters, unroll_length, pack, callbacks)
+        for o in outputs:
+            yield o
+
+        # Do batched policy eval
+        eval_results = _do_policy_eval(tf_sess, to_eval, policies,
+                                       active_episodes)
+
+        # Process results and update episode state
+        actions_to_send = _process_policy_eval_results(
+            to_eval, eval_results, active_episodes, active_envs,
+            off_policy_actions)
 
-        # Map of env_id -> agent_id -> action replies
-        actions_to_send = defaultdict(dict)
+        # Return computed actions to ready envs. We also send to envs that have
+        # taken off-policy actions; those envs are free to ignore the action.
+        async_vector_env.send_actions(actions_to_send)
 
-        # For each environment
-        for env_id, agent_obs in unfiltered_obs.items():
-            new_episode = env_id not in active_episodes
-            episode = active_episodes[env_id]
-            if not new_episode:
-                episode.length += 1
-                episode.batch_builder.count += 1
-                episode._add_agent_rewards(rewards[env_id])
-
-            # Check episode termination conditions
-            if dones[env_id]["__all__"] or episode.length >= horizon:
-                all_done = True
-                atari_metrics = _fetch_atari_metrics(async_vector_env)
-                if atari_metrics is not None:
-                    for m in atari_metrics:
-                        yield m._replace(custom_metrics=episode.custom_metrics)
-                else:
-                    yield RolloutMetrics(episode.length, episode.total_reward,
-                                         dict(episode.agent_rewards),
-                                         episode.custom_metrics)
+
+def _process_observations(async_vector_env, policies, batch_builder_pool,
+                          active_episodes, unfiltered_obs, rewards, dones,
+                          infos, off_policy_actions, horizon, obs_filters,
+                          unroll_length, pack, callbacks):
+    """Record new data from the environment and prepare for policy evaluation.
+
+    Returns:
+        active_envs: set of non-terminated env ids
+        to_eval: map of policy_id to list of agent PolicyEvalData
+        outputs: list of objects to return from the sampler
```

Comment:
this could have a docstring, because it's not immediately clear that this `outputs` could include metrics or Episodes.